### DISCUSSION TITLE
Ideas on re-using internal recipes / automated dev setups

### DISCUSSION THREAD

**Discussion by marvinhinz** [deployphp/deployer#3475] [General]
Hey there,

currently we are using deployer on a central ssh jump server, that has access to all of our webservers. 

Our developers usually dont have direct ssh access to those from their machines, only to the central login server. From there they can deploy by `cd projects/projectname && dep deploy prod`. All the projects deploy.php files and our many custom recipes and deploy flows are centralized in a single git repo. That works pretty well for us.  

Now i am evaluating docker and automatic dev environments. Ideally when a developer checks out an existing project locally, i'd like to have a deploy.php in the git repo root, that contains some internal logic for reusing dockerfile & docker-compose templates and automated project setup commands. 

I really dislike the idea of copy pasting the same docker-compose.yml and reusing it for similar projects. From experience this will lead to configuration drift quickly. 

All that could be solved by using deployer on localhost (dev machine), and some central custom recipes that act as as pre-processor. Maybe there are already solutions for this kind of problem, but i could not find something fitting.

So the question is how can i reuse our custom deployer code on multiple projects. 
Maybe it would be nice if deployer had a function to require external recipes directly, and load them automatically:

```
<?php
namespace Deployer;
recipe('composer-namespace/name');
....

```
Sure this could also be achieved by including the `composer-namespace/name` in the root composer.json, and something like:
 
```
<?php
namespace Deployer;
require __DIR__ . '/vendor/composer-namespace/name/recipe.php';
....

```
But this will lead to file not found errors when for example executing `dep dev:setup` before issuing composer install.
For us this also would have a slight drawback, as we mostly keep the project parts inside separate folders inside the repo, so we currently dont have a composer.json at the repo root. 

Would love to get some opinions on this :)

**Comment by antonmedv** [comment]
I was thinking about C adding ability for Deployer import() functionality to require external recipes via https. 

Will it help?

**Reply by marvinhinz** [reply]
Sounds interesting for public code like raw.github.com/... ,but how can you manage authentication for private code. HTTP Basic Auth inlined in the url? Seems a bit basic / hacky. The composer solution could handle all that automagically in a more sophisticated way:

For example: running `dep XYZ` when a deploy.php is present, that uses import(..) the following logic could be applied:

- check if there is already a composer.json and if  the imports are already in the composer.json require-dev section
- if not, ask the user if he trusts the included repos (show the names) and wants to require them
- if user confirms, add to require-dev section and run composer update for those packages
- check if vendor folder and repos exists, if not either promt the user, or just run composer install
- then proceed with the issued `dep XYZ` command

If you're interested i'd be willing to create a first draft pr
